applications/IgaApplication/custom_elements/shell_3p_element.h

```diff
@@ -0,0 +1,460 @@
+    */
+    void EquationIdVector(
+        EquationIdVectorType& rResult,
+        ProcessInfo& rCurrentProcessInfo
+    ) override;
+
+    /**
+    * @brief Sets on rConditionDofList the degrees of freedom of the considered element geometry
+    * @param rElementalDofList The vector containing the dof of the element
+    * @param rCurrentProcessInfo The current process info instance
+    */
+    void GetDofList(
+        DofsVectorType& rElementalDofList,
+        ProcessInfo& rCurrentProcessInfo
+    ) override;
+
+    ///@}
+    ///@name Base Class Operations
+    ///@{
+
+    void Initialize() override;
+
+    void GetValuesVector(
+        Vector& rValues,
+        int Step) override;
+
+    void GetFirstDerivativesVector(
+        Vector& rValues,
+        int Step) override;
+
+    void GetSecondDerivativesVector(
+        Vector& rValues,
+        int Step) override;
+
+    ///@}
+    ///@name Geometrical Functions
+    ///@{
+
+    void CalculateHessian(
+        Matrix& Hessian,
+        const Matrix& DDN_DDe);
+
+    ///@}
+    ///@name Check
+    ///@{
+
+    /**
+    * This function provides the place to perform checks on the completeness of the input.
+    * It is designed to be called only once (or anyway, not often) typically at the beginning
+    * of the calculations, so to verify that nothing is missing from the input
+    * or that no common error is found.
+    * @param rCurrentProcessInfo
+    */
+    int Check(const ProcessInfo& rCurrentProcessInfo) override;
+
+    ///@}
+    ///@name Input and output
+    ///@{
+
+    /// Turn back information as a string.
+    std::string Info() const override
+    {
+        std::stringstream buffer;
+        buffer << "KLElement #" << Id();
+        return buffer.str();
+    }
+
+    /// Print information about this object.
+    void PrintInfo(std::ostream& rOStream) const override
+    {
+        rOStream << "KLElement #" << Id();
+    }
+
+    /// Print object's data.
+    void PrintData(std::ostream& rOStream) const {
+        pGetGeometry()->PrintData(rOStream);
+    }
+
+    ///@}
+
+private:
+    ///@name Member Variables
+    ///@{
+
+    // Components of the metric coefficient tensor on the contravariant basis
+    std::vector<array_1d<double, 3>> m_A_ab_covariant_vector;
+    // Components of the curvature coefficient tensor on the contravariant basis
+    std::vector<array_1d<double, 3>> m_B_ab_covariant_vector;
+
+    // Determinant of the geometrical Jacobian.
+    Vector m_dA_vector;
+
+    /* Transformation the strain tensor from the curvilinear system
+    *  to the local cartesian in voigt notation including a 2 in the
+    *  shear part. */
+    std::vector<Matrix> m_T_vector;
+
+    /// The vector containing the constitutive laws for all integration points.
+    std::vector<ConstitutiveLaw::Pointer> m_constitutive_law_vector;
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /// Calculates LHS and RHS dependent on flags
+    void CalculateAll(
```

Comment:
Is this necessary?
I know for the solids it makes sense to do it, hence asking

Reply:
It makes the element a bit more structured. For the moment it is good. However, I know that it can be more optimized still. The same function call needs to be done for RightHandSideVector and for entire system.